dlog/plonk/src/index.rs

```diff
@@ -0,0 +1,166 @@
+    }
+}
+
+pub enum SRSSpec <'a, G: CommitmentCurve>{
+    Use(&'a SRS<G>),
+    Generate
+}
+
+impl<'a, G: CommitmentCurve> SRSValue<'a, G> where G::BaseField : PrimeField {
+    pub fn generate(size: usize) -> SRS<G> {
+        SRS::<G>::create(size)
+    }
+
+    pub fn create<'b>(size: usize, spec : SRSSpec<'a, G>) -> SRSValue<'a, G>{
+        match spec {
+            SRSSpec::Use(x) => SRSValue::Ref(x),
+            SRSSpec::Generate => SRSValue::Value(Self::generate(size))
+        }
+    }
+}
+
+pub struct Index<'a, G: CommitmentCurve> where G::ScalarField : QnrField
+{
+    // constraints as Lagrange-based polynoms
+    pub cs: ConstraintSystem<Fr<G>>,
+
+    // polynomial commitment keys
+    pub srs: SRSValue<'a, G>,
+    
+    // maximal size of polynomial section
+    pub max_poly_size: usize,
+
+    // random oracle argument parameters
+    pub fr_sponge_params: ArithmeticSpongeParams<Fr<G>>,
+    pub fq_sponge_params: ArithmeticSpongeParams<Fq<G>>,
+}
+
+pub struct VerifierIndex<'a, G: CommitmentCurve>
+{
+    pub domain: D<Fr<G>>,          // evaluation domain
+    pub max_poly_size: usize,           // maximal size of polynomial section
+    pub srs: SRSValue<'a, G>,           // polynomial commitment keys
+
+    // index polynomial commitments
+    pub sigma_comm: [PolyComm<G>; 3],   // permutation commitment array
+    pub sid_comm:   PolyComm<G>,        // SID commitment
+    pub ql_comm:    PolyComm<G>,        // left input wire commitment
+    pub qr_comm:    PolyComm<G>,        // right input wire commitment
+    pub qo_comm:    PolyComm<G>,        // output wire commitment
+    pub qm_comm:    PolyComm<G>,        // multiplication commitment
+    pub qc_comm:    PolyComm<G>,        // constant wire commitment
+
+    // poseidon polynomial commitments
+    pub rcm_comm:   [PolyComm<G>; 3],   // round constant polynomia commitment array
+    pub fpm_comm:   PolyComm<G>,        // full/partial round indicator polynomial commitment
+    pub pfm_comm:   PolyComm<G>,        // partial/full round indicator polynomial commitment
+    pub psm_comm:   PolyComm<G>,        // poseidon constraint selector polynomialcommitment
+
+    // ECC arithmetic polynomial commitments
+    pub add_comm:   PolyComm<G>,        // EC addition selector polynomial commitment
+    pub mul1_comm:  PolyComm<G>,        // EC variable base scalar multiplication selector polynomial commitment
+    pub mul2_comm:  PolyComm<G>,        // EC variable base scalar multiplication selector polynomial commitment
+    pub emul1_comm: PolyComm<G>,        // endoscalar multiplication selector polynomial commitment
+    pub emul2_comm: PolyComm<G>,        // endoscalar multiplication selector polynomial commitment
+    pub emul3_comm: PolyComm<G>,        // endoscalar multiplication selector polynomial commitment
+
+    pub r:          Fr<G>,              // coordinate shift for right wires
+    pub o:          Fr<G>,              // coordinate shift for output wires
```

Comment:
Does this depends only on the domain? We will want to factor it out on the OCaml side if so.

Reply:
Yes, these values depend on the domain in the sense that they are sampled outside of it.